File: app/src/main/java/org/scop/com/zenloops/GamePanel.java
package org.scop.com.zenloops;

import android.content.Context;
import android.content.ContextWrapper;
import android.graphics.Canvas;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.view.View;

import org.scop.com.zenloops.core.Animate;
import org.scop.com.zenloops.core.Animator;
import org.scop.com.zenloops.elements.Grid;
import org.scop.com.zenloops.elements.Link;
import org.scop.com.zenloops.elements.LinkGraphics;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * Created by Oscar on 18/09/2016.
 */
public class GamePanel extends View implements Animate {
    private Grid grid;
    private ScaleGestureDetector scaleDetector;
    private GestureDetector gestureDetector;
    private String save_state_path;
    public Animator animator;

    public GamePanel(Context context, boolean load) {
        super(context);

        this.save_state_path = new ContextWrapper(context).getFilesDir().getPath()+"/state.save";

        LinkGraphics.getInstance().loadGraphics(this.getContext());
        if (load){
            load = loadState();
        }
        if (!load){
            grid = new Grid(this,10*7/3,21*7/3,0.76f,0.105f);
            grid.randomFill();
        }
        this.scaleDetector = new ScaleGestureDetector(context, new ScaleListener());
        this.scaleDetector.setQuickScaleEnabled(false);
        this.gestureDetector = new GestureDetector(context, new GestureListener());
        this.animator = Animator.getInstance(this);
    }

    private int backgroundColor = 0xFF000000;
    private boolean finishedNotified = false;
    @Override
    public void draw(Canvas canvas) {
        super.draw(canvas);
        if (grid.isFinished() != finishedNotified){
            finishedNotified = !finishedNotified;
            animator.addAnimation(this);
        }
        canvas.drawColor(backgroundColor);
        grid.draw(canvas, x, y, scale, wScaled, hScaled);
    }

    @Override
    public boolean updateAnimation(){
        if (grid.isFinished()){
            backgroundColor+= 0x000F0F0F;
            if (backgroundColor > 0xFFFFFFFF){
                backgroundColor = 0xFFFFFFFF;
                return false;
            }
        } else {
            backgroundColor-= 0x00161616;
            if (backgroundColor < 0xFF000000){
                backgroundColor = 0xFF000000;
                return false;
            }
        }
        return true;
    }

    public void saveState(){
        try {
            new File(save_state_path.substring(0,save_state_path.lastIndexOf("/"))).mkdirs();
            File f = new File(save_state_path);
            if (f.exists()) f.delete();
            FileOutputStream fos = new FileOutputStream (f);
            DataOutputStream dos = new DataOutputStream(fos);

            String str = grid.toPack();
            str+=":"+(int)-x+":"+(int)-y+":"+(int)(scale*100)+"|";

            char[] map = str.toCharArray();
            for (char c : map)
                dos.writeChar(c);
            dos.close();
            fos.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public boolean loadState(){
        try {
            File f = new File(save_state_path);
            FileInputStream fis = new FileInputStream (f);
            DataInputStream dis = new DataInputStream(fis);

            String str = "";
            while (true) {
                char c = dis.readChar();
                if (c=='|') break;
                str += c;
            }

            dis.close();
            fis.close();

            String[] read = str.split(":");

            grid = new Grid(this,Integer.parseInt(read[0]),Integer.parseInt(read[1]),Integer.parseInt(read[2])/10000f,Integer.parseInt(read[3])/10000f);
            grid.fromPack(read[4]);

            x = -Integer.parseInt(read[5]);
            y = -Integer.parseInt(read[6]);
            scale = Integer.parseInt(read[7])/100f;
            wScaled = w/scale;
            hScaled = h/scale;
            postInvalidate();
            return true;
        } catch (EOFException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    // CONTROLS:
    private boolean isResizing = false;
    private boolean isMoving = false;
    private float w,h,x,y,wScaled,hScaled,scale=0;
    private float minScale = 0.3f;
    private float maxScale = 2f;
    private float posX0,posY0;

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        this.w = w;
        this.h = h;
        int[] vp = grid.getViewport();

        this.minScale = Math.min( (float)w/Math.min(vp[0],3164), (float)h/Math.min(vp[1],5780))*0.94f;
        this.maxScale = Math.min( (float)w/LinkGraphics.WIDTH/3, (float)h/LinkGraphics.HEIGHT/3);
        if (this.scale==0) this.scale = (this.minScale*3+this.maxScale)/4;

        this.hScaled = h/scale;
        this.wScaled = w/scale;
        super.onSizeChanged(w, h, oldw, oldh);
    }

    public boolean onTouchEvent(MotionEvent e) {
        gestureDetector.onTouchEvent(e);
        scaleDetector.onTouchEvent(e);

        if (!scaleDetector.isInProgress()) {
            int action = e.getActionMasked();
            switch (action) {
                case MotionEvent.ACTION_DOWN:
                    posX0 = e.getX();
                    posY0 = e.getY();
                    break;

                case MotionEvent.ACTION_MOVE:
                    if (this.isResizing) return true;

                    float X = e.getX();
                    float Y = e.getY();

                    int dx = Math.round((X - posX0) / scale);
                    int dy = Math.round((Y - posY0) / scale);

                    if (isMoving || Math.abs(dx)+Math.abs(dy)>35){
                        posX0 = X;
                        posY0 = Y;
                        x+=dx;
                        y+=dy;
                        adjustXY();
                        this.isMoving = true;
                        postInvalidate();
                    }
                    break;

                case MotionEvent.ACTION_OUTSIDE:
                case MotionEvent.ACTION_UP:
                    this.isMoving = false;
                    this.isResizing = false;
                    break;
                case MotionEvent.ACTION_CANCEL:
                    break;
            }
        }
        return true;//super.onTouchEvent(event);
    }

    private void adjustXY(){
        int[] vp = grid.getViewport();

        float halfWScaled = wScaled/2,
                halfHScaled = hScaled/2;
        vp[0] += halfWScaled;
        vp[1] += halfHScaled;

        //halfWScaled = Math.max(halfWScaled,2000);
        if (x>halfWScaled) x=halfWScaled;
        if (y>halfHScaled) y=halfHScaled;

        float iw = x-wScaled;
        float ih = y-hScaled;

        if (-iw>vp[0]) x = wScaled-vp[0];
        if (-ih>vp[1]) y = hScaled-vp[1];
    }

    private class ScaleListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {
        @Override
        public boolean onScale(ScaleGestureDetector detector) {
            isResizing = true;

            float ratio = detector.getScaleFactor();

            float iScale = scale;
            scale *= ratio;
            scale = Math.max(minScale, Math.min(scale, maxScale));
            wScaled = w/scale;
            hScaled = h/scale;

            float dd = (1/scale - 1/iScale);
            x += dd*w/2;
            y += dd*h/2;

            postInvalidate();
            return true;
        }
    }
    private class GestureListener extends GestureDetector.SimpleOnGestureListener {
        @Override
        public void onLongPress(MotionEvent e) {
            //if (grid.isFinished()){
            if (!isResizing && !isMoving){
                grid.restartNew();
                postInvalidate();
            }
        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            GamePanel p = GamePanel.this;
            float x = e.getX()/p.scale-p.x;
            float y = e.getY()/p.scale-p.y;
            boolean success = grid.rotate(x,y);
            if (success){
                grid.validate();
                postInvalidate();
            }
            return true;
        }
        @Override
        public boolean onSingleTapConfirmed(MotionEvent e){
            return true;
        }
        @Override
        public boolean onDoubleTapEvent(MotionEvent e) {
            if (e.getAction()!=MotionEvent.ACTION_DOWN) isMoving = true;
            if (e.getAction()!=MotionEvent.ACTION_UP) return false;
            onSingleTapUp(e);
            return true;
        }
    }
}

File: app/src/main/java/org/oar/zenloops/GamePanel.kt
package org.oar.zenloops

import android.animation.ArgbEvaluator
import android.annotation.SuppressLint
import android.content.Context
import android.content.ContextWrapper
import android.graphics.Canvas
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.ScaleGestureDetector.SimpleOnScaleGestureListener
import androidx.core.content.ContextCompat
import org.oar.zenloops.core.Animate
import org.oar.zenloops.core.ViewAnimator
import org.oar.zenloops.elements.Grid
import org.oar.zenloops.elements.LinkGraphics
import java.io.*
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

@SuppressLint("ViewConstructor")
class GamePanel(
    context: Context,
    load: Boolean
) : ViewAnimator(context), Animate {

    private var grid: Grid
    private val localFileSaveState: String

    private val gestureDetector = GestureDetector(context, GestureListener())
    private val scaleDetector = ScaleGestureDetector(context, ScaleListener())
        .apply { isQuickScaleEnabled = false }

    // BACKGROUND & ANIMATION
    private val bgColor = ContextCompat.getColor(context, R.color.backgroundColor)
    private val endBgColor = ContextCompat.getColor(context, R.color.endBackgroundColor)
    private var currentBgColor = bgColor
    private val argbEvaluator = ArgbEvaluator()
    private var animationStep = 0f
    private var finished = false

    // CONTROLS:
    private var isResizing = false
    private var isMoving = false
    private var width = 0f
    private var height = 0f
    private var xPos = 0f
    private var yPos = 0f
    private var wScaled = 0f
    private var hScaled = 0f
    private var zoom = 0f
    private var minZoom = 0.3f
    private var maxZoom = 2f

    private var initialTouchedPositionX = 0f
    private var initialTouchedPositionY = 0f

    init {
        localFileSaveState = ContextWrapper(context).filesDir.path + "/state.save"
        LinkGraphics.loadGraphics(context)

        val loadedGrid = if (load) loadState() else null
        grid = loadedGrid ?: Grid(this, 16, 35, 0.7f).apply { randomFill() }
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)
        if (grid.isFinished != finished) {
            finished = !finished
            addAnimation(this)
        }

        canvas.drawColor(currentBgColor)
        grid.draw(canvas, xPos, yPos, zoom, wScaled, hScaled)
    }

    override fun updateAnimation(): Boolean {
        animationStep = min(animationStep + 0.15f, 1f)

        currentBgColor = if (grid.isFinished)
            argbEvaluator.evaluate(animationStep, bgColor, endBgColor) as Int
        else
            argbEvaluator.evaluate(animationStep, endBgColor, bgColor) as Int

        if (animationStep >= 1) {
            animationStep = 0f
            return false
        }
        return true
    }

    fun saveState() {
        try {
            File(localFileSaveState.substring(0, localFileSaveState.lastIndexOf("/"))).mkdirs()
            val f = File(localFileSaveState)
            if (f.exists()) f.delete()
            val fos = FileOutputStream(f)
            val dos = DataOutputStream(fos)
            var str = grid.toPack()
            str += ":" + -xPos.toInt() + ":" + -yPos.toInt() + ":" + (zoom * 100).toInt() + "|"
            val map = str.toCharArray()
            for (c in map) dos.writeChar(c.code)
            dos.close()
            fos.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun loadState(): Grid? {
        try {
            val f = File(localFileSaveState)
            val fis = FileInputStream(f)
            val dis = DataInputStream(fis)
            var str = ""
            while (true) {
                val c = dis.readChar()
                if (c == '|') break
                str += c
            }
            dis.close()
            fis.close()
            val read = str.split(":").toTypedArray()
            val grid = Grid(
                this,
                read[0].toInt(),
                read[1].toInt(),
                read[2].toInt() / 10000f,
                read[3].toInt() / 10000f
            )
            grid.fromPack(read[4])
            xPos = -read[5].toInt().toFloat()
            yPos = -read[6].toInt().toFloat()
            zoom = read[7].toInt() / 100f
            wScaled = width / zoom
            hScaled = height / zoom
            postInvalidate()
            return grid
        } catch (e: EOFException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        this.width = w.toFloat()
        this.height = h.toFloat()
        val viewPort = grid.viewport

        minZoom = min(
            this.width / min(viewPort.width, 3164),
            this.height / min(viewPort.height, 5780)) * 0.64f

        maxZoom = min(
            this.width / LinkGraphics.WIDTH / 3,
            this.height / LinkGraphics.HEIGHT / 3)

        if (zoom == 0f) { // defaultZoom
            zoom = (minZoom * 3 + maxZoom) / 4
        }
        hScaled = h / zoom
        wScaled = w / zoom
        super.onSizeChanged(w, h, oldw, oldh)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        gestureDetector.onTouchEvent(event)
        scaleDetector.onTouchEvent(event)
        if (!scaleDetector.isInProgress) {
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    initialTouchedPositionX = event.x
                    initialTouchedPositionY = event.y
                }
                MotionEvent.ACTION_MOVE -> {
                    if (isResizing) return true

                    val dx = ((event.x - initialTouchedPositionX) / zoom).roundToInt()
                    val dy = ((event.y - initialTouchedPositionY) / zoom).roundToInt()

                    if (isMoving || abs(dx) + abs(dy) > 35) {
                        initialTouchedPositionX = event.x
                        initialTouchedPositionY = event.y
                        xPos += dx.toFloat()
                        yPos += dy.toFloat()
                        adjustXY()
                        isMoving = true
                        postInvalidate()
                    }
                }
                MotionEvent.ACTION_OUTSIDE, MotionEvent.ACTION_UP -> {
                    isMoving = false
                    isResizing = false
                }
                MotionEvent.ACTION_CANCEL -> {}
            }
        }
        return true //super.onTouchEvent(event);
    }

    private fun adjustXY() {
        val viewPort = grid.viewport
        val halfWScaled = wScaled / 2
        val halfHScaled = hScaled / 2
        viewPort.width += halfWScaled.toInt()
        viewPort.height += halfHScaled.toInt()

        //halfWScaled = Math.max(halfWScaled,2000);
        if (xPos > halfWScaled) xPos = halfWScaled
        if (yPos > halfHScaled) yPos = halfHScaled

        val iw = wScaled - xPos
        val ih = hScaled - yPos

        if (iw > viewPort.width)  xPos = wScaled - viewPort.width
        if (ih > viewPort.height) yPos = hScaled - viewPort.height
    }

    private inner class ScaleListener : SimpleOnScaleGestureListener() {
        override fun onScale(detector: ScaleGestureDetector): Boolean {
            isResizing = true
            val ratio = detector.scaleFactor
            val iScale = zoom

            zoom *= ratio
            zoom = max(minZoom, min(zoom, maxZoom))

            wScaled = width / zoom
            hScaled = height / zoom

            val dd = 1 / zoom - 1 / iScale
            xPos += dd * width / 2
            yPos += dd * height / 2

            postInvalidate()
            return true
        }
    }

    private inner class GestureListener : SimpleOnGestureListener() {
        override fun onLongPress(event: MotionEvent) {
            //if (grid.isFinished()){
            if (!isResizing && !isMoving) {
                grid.restartNew()
                postInvalidate()
            }
        }

        override fun onSingleTapUp(event: MotionEvent): Boolean {
            val x = event.x / zoom - xPos
            val y = event.y / zoom - yPos

            val success = grid.rotate(x, y)
            if (success) {
                grid.validate()
                postInvalidate()
            }
            return true
        }

        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            return true
        }

        override fun onDoubleTapEvent(e: MotionEvent): Boolean {
            if (e.action != MotionEvent.ACTION_DOWN) isMoving = true
            if (e.action != MotionEvent.ACTION_UP) return false
            onSingleTapUp(e)
            return true
        }
    }
}